### PR TITLE
Use dedicated attributes to fill proxy authentication

### LINKED ISSUE #2512: Proxy Authentication with support for special characters in password
**Describe the bug**
Proxy password can contain special characters such as @,# so on. Current proxy with authentication setting in Insomnia doesn't support special characters.

**To Reproduce**
1. Go to 'preference -> http network proxy '
2. Put on url as 'http://user:p@ss#@proxyurlhost:port
3. Do a normal API request
4. See error 'couldn't resolve proxy name'

**Expected behavior**
Proxy with authentication should work with any password.

**Desktop **
 - OS: [macOS 10.14.6]
 - App Version [2020.3.3]

### PR DESCRIPTION
**Pull request by jgiovaresco**
This PR adds 2 attributes in `Settings` to keep the proxy authentication.

We then provide the user and password to cUrl using the dedicated options: `PROXYUSERNAME` and `PROXYPASSWORD`.

Closes #2512

### CODE REVIEW

**Comment by jgiovaresco** [comment]
@develohpanda 

I think I should be able to add tests in [network.test.js](https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/network/__tests__/network.test.js) but I'm not sure to understand: is there a real call done or is there a mock/stub somewhere 🤔 

I think I should update [axios helper](https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/network/axios-request.js) to handle the HTTP proxy authentication. However, it seems that proxy authentication is (deliberately?) skipped. Should I make the change?

File: packages/insomnia-app/app/network/network.js [hunk 1]
```diff
@@ -479,13 +479,15 @@ export async function _actuallySend(
...
-        const { httpProxy, httpsProxy, noProxy } = settings;
+        const { httpProxy, httpsProxy, noProxy, httpProxyUser, httpProxyPassword } = settings;
         const proxyHost = protocol === 'https:' ? httpsProxy : httpProxy;
         const proxy = proxyHost ? setDefaultProtocol(proxyHost) : null;
         addTimelineText(`Enable network proxy for ${protocol || ''}`);
         if (proxy) {
           setOpt(Curl.option.PROXY, proxy);
           setOpt(Curl.option.PROXYAUTH, CurlAuth.Any);
+          setOpt(Curl.option.PROXYUSERNAME, httpProxyUser);
+          setOpt(Curl.option.PROXYPASSWORD, httpProxyPassword);
```
Comment (develohpanda) [inline packages/insomnia-app/app/network/network.js]: I think we should conditionally set this only if the user and pass are provided.

How will libcurl behave for existing users where username and password (without special chars) is provided as part of the url. In that case, this PR will set the PROXYUSERNAME and PROXYPASSWORD options to `''`; could that override and break proxy auth for current users?

**Review by develohpanda — COMMENTED** [review]
> I think I should be able to add tests in [network.test.js](https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/network/__tests__/network.test.js?rgh-link-date=2020-12-06T17%3A17%3A13Z) but I'm not sure to understand: is there a real call done or is there a mock/stub somewhere 🤔

`node-libcurl` is mocked automatically by jest using this file https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/__mocks__/node-libcurl.js, so no request is actually being sent.
 
> I think I should update [axios helper](https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/network/axios-request.js?rgh-link-date=2020-12-06T17%3A17%3A13Z) to handle the HTTP proxy authentication. However, it seems that proxy authentication is (deliberately?) skipped. Should I make the change?

If it isn't crucial for this PR, I suggest not making that change here. As far as I know the axios helper is used (at least) by isomorphic-git which provides the git sync functionality in Designer. There has been talk on GH I think about proxy auth for git, so it might be worth addressing, but separately IMO.

File: packages/insomnia-app/app/ui/components/settings/general.js [hunk 2]
```diff
@@ -441,6 +441,16 @@ class General extends React.PureComponent<Props, State> {
...
         </div>
+        <div className="form-row pad-top-sm">
+          {this.renderTextSetting('HTTP Proxy User', 'httpProxyUser', '', {
+            placeholder: 'user',
+            disabled: !settings.proxyEnabled,
+          })}
+          {this.renderTextSetting('HTTPS Proxy Password', 'httpProxyPassword', '', {
+            placeholder: 'pass',
+            disabled: !settings.proxyEnabled,
+          })}
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/components/settings/general.js]: The password field should probably apply `type="password"` on the the underlying `input` element.

File: packages/insomnia-app/app/network/network.js (hunk 1 above)
Comment (jgiovaresco) [reply packages/insomnia-app/app/network/network.js]: > How will libcurl behave for existing users where username and password (without special chars) is provided as part of the url. In that case, this PR will set the PROXYUSERNAME and PROXYPASSWORD options to ''; could that override and break proxy auth for current users?

That's a good point... I will dig into this.

File: packages/insomnia-app/app/ui/components/settings/general.js (hunk 2 above)
Comment (jgiovaresco) [reply packages/insomnia-app/app/ui/components/settings/general.js]: Yes, it should. 
I was trying a quick implementation to validate the overall behavior. I was planning to do it before making the PR ready for review. 😅

**Comment by jgiovaresco** [comment]
@develohpanda 
Concerning the axios-helper, it is also exported in
https://github.com/Kong/insomnia/blob/8b278987872668665eb016449a5268300f63cbfd/packages/insomnia-app/app/plugins/context/app.js#L117L120
but it seems this is a private "export".

So I think we don't need to make the change here.

**Comment by jackkav** [comment]
Closed as stale.

**Pull request closed without merging**